compair/impersonation.py

```diff
@@ -0,0 +1,156 @@
+        self._impersonation_overlay = None
+
+    def init_app(self, app):
+        app.before_request(self._process_request)
+        app.after_request(self._process_response)
+        # app.teardown_request(self._teardown)
+
+    def user_loader(self, callback):
+        '''
+        Inject the callback for loading a user. Should be same as the user loader
+        used for flask-login
+        '''
+        self._user_callback = callback
+        return callback
+
+    def authorize(self, callback):
+        '''
+        Inject the callback for authorizing impersonation
+        '''
+        self._impersonation_authorize = callback
+        return callback
+
+    def visual_overlay(self, callback):
+        '''
+        Inject the callback for showing visual cue to html response
+        '''
+        self._impersonation_overlay = callback
+        return callback
+
+    def start_impersonation(self, act_as_user_id):
+        '''
+        Verify logged in user can do impersonation and switch the current_user of flask-login
+        '''
+        if self._can_impersonate(act_as_user_id):
+            # store the current login user info to switch back later
+            session['impersonation_original_user_id'] = current_user.get_id()
+            session['impersonation_act_as_user_id'] = act_as_user_id
+
+            login_user(self._load_user(act_as_user_id))
+
+            return act_as_user_id
+
+        return None
+
+    def end_impersonation(self):
+        '''
+        End the impersonation.
+        '''
+        orig_user_id = self.get_impersonation_original_user_id()
+        act_as_user_id = self.get_impersonation_act_as_user_id()
+        if orig_user_id is not None and act_as_user_id is not None:
+            login_user(self._load_user(orig_user_id))
+
+        # clean up
+        for key in self._SESSION_KEYS:
+            session.pop(key, None)
+
+    def get_impersonation_act_as_user_id(self):
+        '''
+        Check if impersonating another user.  If so, return that user id.  Otherwise, return None
+        '''
+        return session.get('impersonation_act_as_user_id', None)
+
+    def get_impersonation_original_user_id(self):
+        '''
+        Check if impersonating another user.  If so, return user id of original user.  Otherwise, return None
+        '''
+        return session.get('impersonation_original_user_id', None)
+
+    def is_impersonating(self):
+        return self.get_impersonation_act_as_user_id() is not None and \
+               self.get_impersonation_original_user_id() is not None
+
+    def _process_request(self):
+        if self.get_impersonation_act_as_user_id() is None:
+            return None     # proceed with the request
+
+        if current_user and current_user.is_authenticated and not current_user.is_anonymous:
+            if (request.method not in self._IMPERSONATE_ALLOWED_METHODS):
+                flask_abort(403, 'Function not allowed during impersonation')
```

Comment:
My original idea was to keep `impersonation.py` independent from compPAIR.  But since we will need to allow `DELETE` requests for stopping impersonation, I will move the logic into comPAIR and inject it into `impersonation.py` as callback.